### DISCUSSION TITLE
Can I know what key a player pressed?

### DISCUSSION THREAD

**Discussion by KilinW** [saul/demofile#292] [Questions] [answered]
Can I know what key a player pressed?
More precisely what movement command a player use like +forward +moveleft +jump +attack and so.
Originally, I try to get these information from class Player. But I can't see them anywhere.
And I saw some hope in Interface IUserCmd.
But I'm extremely inexperienced about javascript. I don't know what is Interface and how to access it.
Seems like Interface define how to interact with Classes. 
Can someone help me?

**Comment by saul** [comment, marked as answer]
Yes, this is possible, but only for perspective/POV demos. This information isn't recorded (directly) in GOTV demos. Find some example TypeScript below:

```ts
import * as assert from "assert";
import { DemoFile } from "demofile";
import * as fs from "fs";

function parseDemoFile(path: string) {
  fs.readFile(path, (err, buffer) => {
    assert.ifError(err);

    const demoFile = new DemoFile();

    demoFile.on("usercmd", usercmd => {
      // Only log every 1000th user command to reduce spam
      if (usercmd.tickCount % 1000 !== 0) return;

      console.log("usercmd", usercmd.buttons);
    });

    demoFile.on("end", e => {
      if (e.error) {
        console.error("Error during parsing:", e.error);
        process.exitCode = 1;
      }

      console.log("Finished.");
    });

    // Start parsing the buffer now that we've added our event listeners
    demoFile.parse(buffer);
  });
}

parseDemoFile(process.argv[2]);
```

It will produce output similar to:

```
usercmd [ 'forward' ]
usercmd [ 'back', 'moveright' ]
usercmd []
usercmd [ 'forward', 'moveleft' ]
usercmd [ 'forward', 'moveleft' ]
usercmd []
Finished.
```

**Comment by BlackYuzia** [comment]
@saul hmm, that didn't work.
I test with `demo.emit("usercmd", "test")` and my event was trigger.
```js
demo.on("usercmd", (e) => console.log(e)); // didn't trigger without manual emit("usercmd", ...)
```

**Reply by saul** [reply]
See in my answer above: "this is possible, but only for perspective/POV demos. This information isn't recorded (directly) in GOTV demos."
